addr_manager_ds.go

```diff
@@ -0,0 +1,316 @@
+	subsManager *AddrSubManager
+}
+
+// NewDatastoreAddrManager initializes a new DatastoreAddrManager given a
+// Datastore instance, a context for managing the TTL manager, and the interval
+// at which the TTL manager should sweep the Datastore.
+func NewDatastoreAddrManager(ctx context.Context, ds ds.Batching, ttlInterval time.Duration) *DatastoreAddrManager {
+	mgr := &DatastoreAddrManager{
+		ds:          ds,
+		ttlManager:  newTTLManager(ctx, ds, ttlInterval),
+		subsManager: NewAddrSubManager(),
+	}
+	return mgr
+}
+
+// Stop will signal the TTL manager to stop and block until it returns.
+func (mgr *DatastoreAddrManager) Stop() {
+	mgr.ttlManager.cancel()
+}
+
+func peerAddressKey(p *peer.ID, addr *ma.Multiaddr) (ds.Key, error) {
+	hash, err := mh.Sum((*addr).Bytes(), mh.MURMUR3, -1)
+	if err != nil {
+		return ds.Key{}, nil
+	}
+	return ds.NewKey(peer.IDB58Encode(*p)).ChildString(hash.B58String()), nil
+}
+
+func peerIDFromKey(key ds.Key) (peer.ID, error) {
+	idstring := key.Parent().Name()
+	return peer.IDB58Decode(idstring)
+}
+
+// AddAddr will add a new address if it's not already in the AddrBook.
+func (mgr *DatastoreAddrManager) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+// AddAddrs will add many new addresses if they're not already in the AddrBook.
+func (mgr *DatastoreAddrManager) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	if ttl <= 0 {
+		return
+	}
+
+	mgr.setAddrs(p, addrs, ttl, true)
+}
+
+// SetAddr will add or update the TTL of an address in the AddrBook.
+func (mgr *DatastoreAddrManager) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.SetAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+// SetAddrs will add or update the TTLs of addresses in the AddrBook.
+func (mgr *DatastoreAddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	mgr.setAddrs(p, addrs, ttl, false)
+}
+
+func (mgr *DatastoreAddrManager) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, add bool) {
+	for i := 0; i < dsWriteRetries; i++ {
+		// keys to add to the TTL manager
+		var keys []ds.Key
+		batch, err := mgr.ds.Batch()
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		for _, addr := range addrs {
+			if addr == nil {
+				continue
+			}
+
+			key, err := peerAddressKey(&p, &addr)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			keys = append(keys, key)
+
+			if ttl <= 0 {
+				batch.Delete(key)
+				continue
+			}
+			has, err := mgr.ds.Has(key)
+			if err != nil || !has {
+				mgr.subsManager.BroadcastAddr(p, addr)
+			}
+
+			// Allows us to support AddAddr and SetAddr in one function
+			if !has || !add {
+				if err := batch.Put(key, addr.Bytes()); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+		if err := batch.Commit(); err != nil {
+			log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+			continue
+		}
+		mgr.ttlManager.setTTLs(keys, ttl)
```

Comment:
i'd expect the 'add' flag to be passed into here. the intent of `AddAddr` vs `SetAddr` is to only extend the ttl, where set overwrites it.

So if the current ttl is 10, 'add(5)' has no effect, while 'set(5)' results in the ttl being changed to 5. both 'add(12)' and 'set(12)' result in the ttl being changed to 12.

Reply:
yep, fixed!